File: constants/coordinates.py
import requests
from geopy.distance import geodesic

import addresses
import coordinates_and_radius

def get_map_of_coordinates():

     # Get a map of location addresses
    map_of_addresses = addresses.get_map_of_addresses()
    
    # Dictionary to store the updated results
    map_of_coordinates_and_radius = {}

    # Replace the values from addresses to coordinates and radius
    for address, location in map_of_addresses.items():
        result = get_coordinates_and_radius(address)
        if result:  # If the result is not None
            map_of_coordinates_and_radius[address] = result

    # You can return this dictionary or process it further
    return map_of_coordinates_and_radius


def get_coordinates_and_radius(address):

    headers = {
    "User-Agent": "MyGeocodingApp/1.0 (contact@mydomain.com)"
}

    # Nominatim API URL
    url = f'https://nominatim.openstreetmap.org/search?q={address}&format=json&addressdetails=1'


    # Send the request to the Nominatim API
    response = requests.get(url, headers=headers)

    # Attempts to read the data from api
    if response.status_code == 200:
        try:
            data = response.json()
        except ValueError as e:
            print(f"Error parsing JSON: {e}")
            return None
    else:
        print(f"Error: Received status code {response.status_code}")
        return None

    # If a result is found, extract latitude, longitude, and bounding box
    if data:

        # Latitude and longitude
        lat = float(data[0]['lat'])
        lon = float(data[0]['lon'])

        # Bounding box (min_lat, min_lon, max_lat, max_lon)
        min_lat = float(data[0]['boundingbox'][0])
        max_lat = float(data[0]['boundingbox'][1])
        min_lon = float(data[0]['boundingbox'][2])
        max_lon = float(data[0]['boundingbox'][3])

        # Calculate the diagonal distance to estimate the radius
        # We'll use the diagonal of the bounding box as an approximate measure for the radius.
        
        # Get two opposite corners of the bounding box
        point1 = (min_lat, min_lon)
        point2 = (max_lat, max_lon)

        # Calculate the distance between the two corners (diagonal distance)
        distance = geodesic(point1, point2).kilometers

        # Estimated radius as half the diagonal distance (rough estimate)
        radius = distance / 2

        # Creates a dictionary with coordinates and radius
        map_of_coordinates_and_radius = coordinates_and_radius.get_map_of_coordinates_and_radius(lat, lon, radius)
        
        return map_of_coordinates_and_radius
    
    else:

        return None

File: constants/coordinates.ts
import axios from 'axios';
import * as addresses from './addresses';
import * as coordinatesAndRadius from './coordinates_and_radius';
import { getDistance } from 'geolib';

interface NominatimResponse {
    lat: string;
    lon: string;
    boundingbox: string[];
}

interface CoordinatesAndRadius {
    coordinates: {
        latitude: number;
        longitude: number;
    };
    radius: number;
}

// Define the type for specificPlace: { [key: string]: string }
interface SpecificPlace {
    [key: string]: string; // Address of the place as a string
}

/** Returns a map of place categories, locations, and the location's coordinates. */
export async function getMapOfCoordinates(): Promise<any> {
    const mapOfAddresses = await addresses.getMapOfAddresses(); // Await this call if it's async

    // Use Promise.all to handle async updates concurrently
    const updates = Object.values(mapOfAddresses).map(async (locations: any) => {
        for (const specificPlace of Object.values(locations)) {
            // Cast `specificPlace` to the correct type (SpecificPlace)
            const specificPlaceTyped = specificPlace as SpecificPlace;
            for (const [place, address] of Object.entries(specificPlaceTyped)) {
                const result = await getCoordinatesAndRadius(address);
                if (result) {
                    // Type assertion ensures 'specificPlace' value is a string
                    specificPlaceTyped[place] = `${result.coordinates.latitude}, ${result.coordinates.longitude}, radius: ${result.radius}`;
                }
            }
        }
    });

    // Wait for all the updates to complete
    await Promise.all(updates);

    return mapOfAddresses;
}

/** Calculates the coordinates and returns a map including them. */
async function getCoordinatesAndRadius(address: string): Promise<CoordinatesAndRadius | null> {
    const headers = {
        'User-Agent': 'MyGeocodingApp/1.0 (contact@mydomain.com)',
    };

    const url = `https://nominatim.openstreetmap.org/search?q=${address}&format=json&addressdetails=1`;

    try {
        const response = await axios.get<NominatimResponse[]>(url, { headers });

        if (response.status === 200 && response.data.length > 0) {
            const data = response.data[0];
            const lat = parseFloat(data.lat);
            const lon = parseFloat(data.lon);

            const [minLat, maxLat, minLon, maxLon] = data.boundingbox.map((item: string) => parseFloat(item));

            // Calculate the diagonal distance to estimate the radius
            const point1 = { latitude: minLat, longitude: minLon };
            const point2 = { latitude: maxLat, longitude: maxLon };
            const distance = getDistance(point1, point2) / 1000; // convert to kilometers
            const radius = distance / 2;

            return coordinatesAndRadius.getMapOfCoordinatesAndRadius(lat, lon, radius);
        }
    } catch (error) {
        console.error(`Error fetching data: ${error}`);
    }

    return null;
}
